### PR TITLE
optimize `inflate::core::transfer`

### PR DESCRIPTION
**Pull request by connorskees**
Profiling the decoding of a sample image using `image-rs/png` showed that >1/3rd of the time decoding was spent in `inflate::core::transfer`. This change special cases the scenario in which we do no masking to significantly optimize this function for real workloads. On the image that inspired this optimization work, I see a 55% speedup locally.

Relevant benchmarks:

```rs
// before
test oxide::decompress_compressed_lvl_1 ... bench:      89,253 ns/iter (+/- 1,412)
test oxide::decompress_compressed_lvl_6 ... bench:     176,299 ns/iter (+/- 6,266)
test oxide::decompress_compressed_lvl_9 ... bench:     175,840 ns/iter (+/- 3,131)
```

```rs
// after
test oxide::decompress_compressed_lvl_1 ... bench:      83,992 ns/iter (+/- 1,347)
test oxide::decompress_compressed_lvl_6 ... bench:      76,640 ns/iter (+/- 2,288)
test oxide::decompress_compressed_lvl_9 ... bench:      76,791 ns/iter (+/- 1,978)
```

Omitted benchmarks do not execute this codepath. For `oxide::decompress_compressed_lvl_6` and `oxide::decompress_compressed_lvl_9` we see incredible improvements of ~2.3x.

#### Potential future work

It should be possible to speed up the case in which `source_pos.abs_diff(out_pos) == 3`, but I wasn't able to find an easy solution that doesn't use unsafe or SIMD intrinsics. The 3 diff case accounts for about 50% of all cases on some benchmarks that do not see a performance increase from this change.

It should also be possible to do much better on the general case when `source_pos.abs_diff(out_pos) >= 4` using either clever hacks to make the compiler do what we want or SIMD intrinsics. I have not profiled how much of an impact a more efficient implementation of this case would have.

### CODE REVIEW

**Comment by oyvindln** [comment]
Really nice.

Yeah lack of safe SIMD is an issue I guess. Often the compiler is smart enough to use it but it varies. rustc and underlying llvm have some options to help debug why it didn't vectorize stuff but it's clunky.

File: miniz_oxide/src/inflate/core.rs [hunk 1]
```diff
@@ -799,13 +799,34 @@ fn transfer(
...
     out_buf_size_mask: usize,
 ) {
-    for _ in 0..match_len >> 2 {
-        out_slice[out_pos] = out_slice[source_pos & out_buf_size_mask];
-        out_slice[out_pos + 1] = out_slice[(source_pos + 1) & out_buf_size_mask];
-        out_slice[out_pos + 2] = out_slice[(source_pos + 2) & out_buf_size_mask];
-        out_slice[out_pos + 3] = out_slice[(source_pos + 3) & out_buf_size_mask];
-        source_pos += 4;
-        out_pos += 4;
+    debug_assert!(out_pos > source_pos);
```
[the remark this answers is not in the document]
Comment (connorskees) [reply miniz_oxide/src/inflate/core.rs:802]: Ah, I wonder why that's the case. I'll update the code to remove the assert and move the check to the conditions.

[the remark this answers is not in the document]
Comment (kmahar) [reply miniz_oxide/src/inflate/core.rs:802]: we're seeing this assertion fail in CI for #mongodb/mongo-rust-driver as well

[the remark this answers is not in the document]
Comment (sfackler) [reply miniz_oxide/src/inflate/core.rs:802]: Could you yank 0.6.3 while waiting for #134 to land?

[the remark this answers is not in the document]
Comment (oyvindln) [reply miniz_oxide/src/inflate/core.rs:802]: Published 0.6.4 with the fix

[the remark this answers is not in the document]
Comment (dburgener) [reply miniz_oxide/src/inflate/core.rs:802]: To confirm, 0.6.4 seems to have addressed the CI failures for me in dburgener/cascade.  Thank you for the quick turnaround!

**Pull request merged**